src/rviz/default_plugin/marker_display.cpp

```diff
@@ -250,9 +250,21 @@ void MarkerDisplay::incomingMarker( const visualization_msgs::Marker::ConstPtr&
 void MarkerDisplay::failedMarker(const ros::MessageEvent<visualization_msgs::Marker>& marker_evt, tf::FilterFailureReason reason)
 {
   visualization_msgs::Marker::ConstPtr marker = marker_evt.getConstMessage();
-  std::string authority = marker_evt.getPublisherName();
-  std::string error = context_->getFrameManager()->discoverFailureReason(marker->header.frame_id, marker->header.stamp, authority, reason);
-  setMarkerStatus(MarkerID(marker->ns, marker->id), StatusProperty::Error, error);
+
+  if (marker->action == visualization_msgs::Marker::DELETE)
+  {
+    processDelete(marker);
+  }
+  else if (marker->action == 3)  // TODO: visualization_msgs::Marker::DELETEALL when message changes in a future version of ROS
+  {
+    deleteAllMarkers();
+  }
+  else
+  {
+    std::string authority = marker_evt.getPublisherName();
+    std::string error = context_->getFrameManager()->discoverFailureReason(marker->header.frame_id, marker->header.stamp, authority, reason);
+    setMarkerStatus(MarkerID(marker->ns, marker->id), StatusProperty::Error, error);
+  }
```

Comment:
That probably is cleaner -- I was just replicating the design I had seen a while ago in MoveIt. I'll update the PR shortly